### PR TITLE
expression: fix the wrong rounding behavior of Decimal

### LINKED ISSUE #32213: Wrong result for cast expr when value exceeds range
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table if exists test.t1;
    create table test.t1(c1 float, c2 double);
    insert into test.t1 values(999.99, 999.99);
    select cast(test.t1.c1 as decimal(4, 1)) from test.t1;

### 2. What did you expect to see? (Required)

    mysql> select cast(test.t1.c1 as decimal(4, 1)) from test.t1;
    +-----------------------------------+
    | cast(test.t1.c1 as decimal(4, 1)) |
    +-----------------------------------+
    |                             999.9 |
    +-----------------------------------+
    1 row in set, 1 warning (0.01 sec)
    
    mysql> show warnings;
    +---------+------+----------------------------------------------------------------------------+
    | Level   | Code | Message                                                                    |
    +---------+------+----------------------------------------------------------------------------+
    | Warning | 1264 | Out of range value for column 'cast(test.t1.c1 as decimal(4, 1))' at row 1 |
    +---------+------+----------------------------------------------------------------------------+
    1 row in set (0.00 sec)
### 3. What did you see instead (Required)

    mysql> select cast(test.t1.c1 as decimal(4, 1)) from test.t1;
    +-----------------------------------+
    | cast(test.t1.c1 as decimal(4, 1)) |
    +-----------------------------------+
    |                            1000.0 |
    +-----------------------------------+
    1 row in set, 1 warning (0.00 sec)
### 4. What is your TiDB version? (Required)

    mysql> select tidb_version();
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | tidb_version()
                              |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | Release Version: v5.5.0-alpha-146-g7fd298d916
    Edition: Community
    Git Commit Hash: 7fd298d916c30ed0060db3581be58b448325819e
    Git Branch: master
    UTC Build Time: 2022-02-08 06:57:26
    GoVersion: go1.16.4
    Race Enabled: false
    TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
    Check Table Before Drop: false |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    1 row in set (0.00 sec)

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: gengliqi <gengliqiii@gmail.com>

### What problem does this PR solve?

Issue Number: close #32213

Problem Summary:

### What is changed and how it works?
See #32213 and https://github.com/pingcap/tidb/issues/32213#issuecomment-1072045193.
Also, I rename the `ModeHalfEven` to `ModeHalfUp` because `ModeHalfEven` is the wrong name for this rounding behavior. (See https://dev.mysql.com/doc/refman/5.7/en/precision-math-rounding.html).

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
expression: fix the wrong rounding behavior of Decimal
```

### CODE REVIEW

**Comment by gengliqi** [comment]
TestStringtoDecimal failed. Seems the `ErrTruncatedWrongVal` should not be removed here.

**Comment by gengliqi** [comment]
> `TestStringtoDecimal` failed. Seems the `ErrTruncatedWrongVal` should not be removed here.

I change it back. PTAL

**Review by guo-shaoge — APPROVED** [review]

File: types/mydecimal.go [hunk 1]
```diff
@@ -52,8 +52,8 @@ const (
 
 	DivFracIncr = 4
 
-	// ModeHalfEven rounds normally.
-	ModeHalfEven RoundMode = 5
+	// ModeHalfUp rounds normally.
+	ModeHalfUp RoundMode = 5
```
Comment (guo-shaoge) [inline types/mydecimal.go:56]: Maybe some comments? Round away from zero if positive, round towards zero if negative.

File: types/datum.go [hunk 2]
```diff
@@ -1429,20 +1429,27 @@ func ProduceDecWithSpecifiedTp(dec *MyDecimal, tp *FieldType, sc *stmtctx.Statem
...
+			*old = *dec
+		}
+		err = dec.Round(dec, decimal, ModeHalfUp)
+		if err != nil {
+			return nil, err
+		}
+
+		if !isZero {
+			_, digitsInt := dec.removeLeadingZeros()
+			if flen-decimal < digitsInt {
```
Comment (XuHuaiyu) [inline types/datum.go:1447]: Add a comment (maybe an example) for why we check this after `Round`?

File: types/datum_test.go [hunk 3]
```diff
@@ -566,3 +567,49 @@ func BenchmarkCompareDatumByReflect(b *testing.B) {
...
 	}
 }
+
+func TestProduceDecWithSpecifiedTp(t *testing.T) {
+	tests := []struct {
+		dec        string
+		flen       int
+		frac       int
+		newDec     string
+		isOverflow bool
```
[the remark this answers is not in the document]
Comment (gengliqi) [reply types/datum_test.go]: Done.

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**